sdk/identity/azure-identity/test/ut/environment_credential.cpp

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/identity/environment_credential.hpp"
+
+#include <azure/core/io/body_stream.hpp>
+#include <azure/core/platform.hpp>
+
+#include "test_transport.hpp"
+
+#include <gtest/gtest.h>
+#include <stdlib.h>
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif
+
+#include <windows.h>
+#endif
+
+#if !defined(WINAPI_PARTITION_DESKTOP) \
+    || WINAPI_PARTITION_DESKTOP // See azure/core/platform.hpp for explanation.
+
+using namespace Azure::Identity;
+
+namespace {
+class EnvironmentOverride {
+  class Environment {
+    static void SetVariable(std::string const& name, std::string const& value)
+    {
+#if defined(_MSC_VER)
+      static_cast<void>(_putenv((name + "=" + value).c_str()));
+#else
+      if (value.empty())
+      {
+        static_cast<void>(unsetenv(name.c_str()));
+      }
+      else
+      {
+        static_cast<void>(setenv(name.c_str(), value.c_str(), 1));
+      }
+#endif
+    }
+
+  public:
+    static std::string GetVariable(std::string const& name)
+    {
+#if defined(_MSC_VER)
+#pragma warning(push)
+// warning C4996: 'getenv': This function or variable may be unsafe. Consider using _dupenv_s
+// instead.
+#pragma warning(disable : 4996)
+#endif
+      auto const result = std::getenv(name.c_str());
+      return result != nullptr ? result : "";
+#if defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+    }
+
+    static void SetVariables(std::map<std::string, std::string> const& vars)
+    {
+      for (auto var : vars)
+      {
+        SetVariable(var.first, var.second);
+      }
+    }
+  };
+
+  std::map<std::string, std::string> m_originalEnv;
+
+public:
+  ~EnvironmentOverride() { Environment::SetVariables(m_originalEnv); }
+
+  EnvironmentOverride(
+      std::string const& tenantId,
+      std::string const& clientId,
+      std::string const& clientSecret,
+      std::string const& authorityHost,
+      std::string const& username,
+      std::string const& password,
+      std::string const& clientCertificatePath)
+  {
+    std::map<std::string, std::string> const NewEnv = {
+        {"AZURE_TENANT_ID", tenantId},
+        {"AZURE_CLIENT_ID", clientId},
+        {"AZURE_CLIENT_SECRET", clientSecret},
+        {"AZURE_AUTHORITY_HOST", authorityHost},
+        {"AZURE_USERNAME", username},
+        {"AZURE_PASSWORD", password},
+        {"AZURE_CLIENT_CERTIFICATE_PATH", clientCertificatePath},
```

Comment:
Correct me if I am mistaken, but this test seems to be overriding some of the commonly used env vars on any contributor developer machine. Is there a way to get the test coverage but use custom env var names rather than things like `AZURE_TENANT_ID`?

I would prefer if running unit tests didn't change my dev machine's env vars.

Reply:
These changes are not machine-wide. They are for the current process only.